src/shims/fs.rs

```diff
@@ -330,68 +411,44 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
         // If the `AT_SYMLINK_NOFOLLOW` flag is set, we query the file's metadata without following
         // symbolic links.
-        let metadata = if flags & this.eval_libc("AT_SYMLINK_NOFOLLOW")?.to_i32()? != 0 {
-            // FIXME: metadata for symlinks need testing.
-            std::fs::symlink_metadata(path)
-        } else {
-            std::fs::metadata(path)
-        };
-
-        let metadata = match metadata {
-            Ok(metadata) => metadata,
-            Err(e) => {
-                this.set_last_error_from_io_error(e)?;
-                return Ok(-1);
-            }
-        };
+        let is_symlink = flags & this.eval_libc("AT_SYMLINK_NOFOLLOW")?.to_i32()? != 0;
 
-        let file_type = metadata.file_type();
-
-        let mode_name = if file_type.is_file() {
-            "S_IFREG"
-        } else if file_type.is_dir() {
-            "S_IFDIR"
-        } else {
-            "S_IFLNK"
+        let status = match FileStatus::new(this, path, is_symlink)? {
+            Some(status) => status,
+            None => return Ok(-1),
         };
 
         // The `mode` field specifies the type of the file and the permissions over the file for
         // the owner, its group and other users. Given that we can only provide the file type
         // without using platform specific methods, we only set the bits corresponding to the file
         // type. This should be an `__u16` but `libc` provides its values as `u32`.
-        let mode: u16 = this
-            .eval_libc(mode_name)?
+        let mode: u16 = status
+            .mode
             .to_u32()?
             .try_into()
-            .unwrap_or_else(|_| bug!("libc contains bad value for `{}` constant", mode_name));
-
-        let size = metadata.len();
+            .unwrap_or_else(|_| bug!("libc contains bad value for constant"));
 
-        let (access_sec, access_nsec) = extract_sec_and_nsec(
-            metadata.accessed(),
-            &mut mask,
-            this.eval_libc("STATX_ATIME")?.to_u32()?,
-        )?;
+        let (access_sec, access_nsec) = status.accessed.map(|tup| {
+            mask |= this.eval_libc("STATX_ATIME")?.to_u32()?;
+            InterpResult::Ok(tup)
+        }).unwrap_or(Ok((0, 0)))?;
```

Comment:
Uh, what is this? Why is it so complicated? And also so imperative, what with adjusting `mask` as a side-effect.

Reply:
That's a: "set the flag if the access time was available, otherwise set the access time to zero". Maybe an `if let` would be better?

Reply:
I see. A comment should be enough, if you'd prefer to keep the code as-is.